src/sys/windows/selector.rs

```diff
@@ -371,14 +411,233 @@ impl Selector {
+            let flags = unsafe {
+                NonZeroU32::new_unchecked(self.flags.get() | rhs.flags.get())
+            };
+            Readiness { flags }
+        }
+    }
+
+    /// Callback trait for completed IOCP operations.
+    ///
+    /// This trait may be implemented to handle completed IOCP operations. The implementing object may
+    /// signal the readiness for certain operations to the upper layer by returning a `Readiness` value
+    /// that is then converted into `mio::Event` and delivered to the user.
+    ///
+    /// This trait is implemented for closures as well.
+    pub trait CompletionCallback {
+        /// Called on completion of the associated operation.
+        ///
+        /// The callback will be called once the operation that it is bound to via a call to
+        /// `Overlapped::new` has been completed and the IOCP is signalled. By returning a `Readiness`
+        /// value, the callback may deliver a `mio::Event` to the user.
+        fn complete_operation(&mut self, entry: &OVERLAPPED_ENTRY) -> Option<Readiness>;
+    }
+
+    impl<F> CompletionCallback for F where F: FnMut(&OVERLAPPED_ENTRY) -> Option<Readiness> {
+        fn complete_operation(&mut self, entry: &OVERLAPPED_ENTRY) -> Option<Readiness> {
+            (*self)(entry)
+        }
+    }
+
+    #[derive(Debug)]
+    pub(crate) struct RawHandleCompletionHandler {
+        token: Token
+    }
+
+    impl IocpHandler for RawHandleCompletionHandler {
+        fn handle_completion(&mut self, status: &CompletionStatus) -> Option<Event> {
+            let overlapped = status.overlapped() as *mut Overlapped;
+            unsafe {
+                debug_assert!((*overlapped).canary == OVERLAPPED_CANARY);
+                (*overlapped).callback.complete_operation(status.entry())
+                    .map(|r| r.into_event(self.token))
+            }
+        }
+    }
+
+    /// A `Binding` is embedded in all I/O objects associated with a `Registry`.
+    ///
+    /// Each registration keeps track of which selector the I/O object is
+    /// associated with, ensuring that implementations of `Source` can be
+    /// conformant for the various methods on Windows.
+    ///
+    /// If you're working with custom IOCP-enabled objects then you'll want to
+    /// ensure that one of these instances is stored in your object and used in the
+    /// implementation of `Source`.
+    ///
+    /// For more information about how to use this see the `windows` module
+    /// documentation in this crate.
+    #[derive(Debug, Clone)]
+    pub struct Binding {
+        selector: Arc<SelectorInner>,
+        token: Token
+    }
+
+    impl Binding {
+        /// Creates a new blank binding ready to be inserted into an I/O
+        /// object.
+        ///
+        /// Won't actually do anything until `register_handle` has been called during a call to
+        /// `Registry::register`.
+        pub fn new(registry: &Registry, token: Token) -> Self {
```

Comment:
Is there a reason why you don't require an `event::Source` on creation of `Binding`?

Reply:
Actually I just need the handle, not the whole source. I split this into two parts as I wanted to give the user a chance to store the `Binding` instance before it starts firing events, so the usual sequence in a driver is:
- Initialize the `Option<Binding>` member of your custom struct with `None`
- Inside `Source::register`, you then set it to `Some(Binding::new(..))`
- *then** give it the handle to work with using `register_handle`